### PR TITLE
Accept additional user-defined classes in fenced code blocks

### LINKED ISSUE #78917: Alternative highlighting for Rust code block in rustdoc
Context https://twitter.com/imperioworld_/status/1325916530257833984 cc @GuillaumeGomez 

# Description

I'm the author of https://github.com/Hywan/inline-c-rs/. It allows to write C code inside Rust. It's a super basic `proc_macro` that transforms the code into a string (with some extra manipulations), which is saved inside a temp file that is compiled by a C/C++ compiler, and finally run. That's useful for testing C API of a Rust program inside Rust with the Rust tooling (like `cargo test`). It is super useful for us at https://github.com/wasmerio/wasmer.

Recently, we pushed the concept further: Using `inline-c-rs` inside documentation, like this:

```rust
/// Blah blah blah.
///
/// # Examples
///
/// ```rust
/// # use inline_c::assert_c;
/// # fn main() {
/// #    (assert_c! {
/// # #include "tests/wasmer_wasm.h"
/// #
/// int main() {
///     wasm_engine_t* engine = wasm_engine_new();
///     wasm_store_t* store = wasm_store_new(engine);
///
///     // etc.
///    
///     return 0;
/// }
/// #    })
/// #    .success();
/// # }
/// ```
#[no_mangle]
pub unsafe extern "C" fn wasm_module_new(/* skipped */) -> /* skipped */ {
    // skipped
}
```

This test is:

* ✅ written in Rust, and contains C,
* ✅ is run by `cargo doc` because the code block is tagged with <code>```rust</code> … so we can test C with `cargo test` 🤪!
* ✅ is compiled to HTML by `rustdoc`.

The result of `cargo doc` looks like this:

Screenshot 2020-11-06 at 11 40 44

That's excellent! All the “Rust decoration” is “removed” (thanks to `# …`), and only the C code stays. That's insane, but it works and that's super fun.

One problem though:

* ❌ The highlighting is incorrect.

Because it's a <code>```rust</code> code block, there is a special handler for that I assume. I tried to write `rust,language-c` but the `language-c` part is ignored and is not present in the generated HTML code, as it can be seen here:

```html
<pre class="rust rust-example-rendered">
```

# Expectation

I would expect <code>```rust</code> to be a special keyword that unfolds to `rust` + `language-rust`:

* The former describes that it's a Rust code block, and consequently is a candidate for a test,
* The second describes the syntax highlighting ruleset to apply for the HTML documentation.

This keyword could have a different meaning in the presence of `language-<lang_id>` which “cancels” `language-rust`.

In other words, writing `rust,language-c` would keep the actual behavior of `rust` but it will disable the default highlighting to allow the user to define another one.

How does it sound?

### PR DESCRIPTION
**Pull request by poliorcetics**
Fix #78917.

This allow users to write the following:

```rust
/// Some code block with `rust,class:test:name` as the language string
///
/// ```rust,class:test:name
/// int main(void) {
///     return 0;
/// }
/// ```
fn main() {}
```

This will notably produce the following HTML:

```html
<pre class="rust test:name">
int main(void) {
    return 0;
}</pre>
```

Instead of:

```html
<pre class="rust rust-example-rendered">
<span class="ident">int</span> <span class="ident">main</span>(<span class="ident">void</span>) {
    <span class="kw">return</span> <span class="number">0</span>;
}
</pre>
```

TODO:

- [x] Implement basic functionality
- [x] Write tests
- [x] Document the feature
- [x] Make the feature nightly only (example PR: #75740)
- [x] Clean up the code a little because my changes are very raw right now
- [x] Clean up the tests, they are starting to get unwieldy (https://github.com/rust-lang/rust/pull/79454#discussion_r531814290)

MERGE:

- [x] wait for #78429

### CODE REVIEW

**Comment by poliorcetics** [comment]
@rustot labels: A-markdown-parsing T-rustdoc

**Comment by poliorcetics** [comment]
@rustot modify labels: A-markdown-parsing T-rustdoc

**Comment by jyn514** [comment]
I'm not sure I understand what this is actually getting us. There's no more rust syntax highlighting, yes - but there's no C source highlighting either, which was kind of the point. And the C highlighting can't use the same mechanism, because rust highlighting uses `rustc_lexer` directly and I do *not* want to tie rustdoc into clang, nor have to do something similar for each language. That means highlight.js or something like it, but that seems somewhat complicated ...

Anyway, I think we should have this feature actually *do* something before we accept the syntax, especially since the new feature is insta-stable currently.

**Comment by jyn514** [comment]
cc @GuillaumeGomez - do you have opinons on highlight.js?

**Comment by Hywan** [comment]
I think it's better to not force a specific Javascript library to highlight the code. One may want highlight.js while another one may prefer prism.js. The primary goal of this PR is to allow passing any HTML class to the code block and to disable the Rust highlighting in this specific case.

**Comment by GuillaumeGomez** [comment]
> I'm not sure I understand what this is actually getting us. There's no more rust syntax highlighting, yes - but there's no C source highlighting either, which was kind of the point. And the C highlighting can't use the same mechanism, because rust highlighting uses `rustc_lexer` directly and I do _not_ want to tie rustdoc into clang, nor have to do something similar for each language. That means highlight.js or something like it, but that seems somewhat complicated ...

That was never the goal. The point here is to allow users to add a class themselves and disable the rustdoc highlighting so that they can then add a JS library which will do it (like highlight.js or prism.js as mentionned by @Hywan).

> Anyway, I think we should have this feature actually _do_ something before we accept the syntax, especially since the new feature is insta-stable currently.

We can make it remain unstable by adding a check when parsing the codeblocks' tags for the moment. I think it would be preferrable in fact. Users will still be able to use it on docs.rs if they want to and that will allow to make breaking changes if needed.

@poliorcetics Can you add the check for nightly and only allow this feature then? (And otherwise, maybe print a warning saying it's nightly only for the moment?)

**Comment by jyn514** [comment]
> That was never the goal. The point here is to allow users to add a class themselves and disable the rustdoc highlighting so that they can then add a JS library which will do it.

Ahhh, that makes way more sense. Ok, I'm on board with this then.

> @poliorcetics Can you add the check for nightly and only allow this feature then? (And otherwise, maybe print a warning saying it's nightly only for the moment?)

This should go through the normal feature gate mechanism, which gives an error when you try to use nightly features on stable.

Another thing is that I think this will change how language strings are parsed, so it should probably wait for https://github.com/rust-lang/rust/pull/78429.

**Comment by poliorcetics** [comment]
> This should go through the normal feature gate mechanism, which gives an error when you try to use nightly features on stable.

I don't know how those work for rustdoc, is there some place can I find the information or an example ?

**Comment by GuillaumeGomez** [comment]
@poliorcetics It's the same as for the rest of the compiler. You can see everything that you need in this [PR](https://github.com/rust-lang/rust/pull/75740) for example (which is removing the feature gate part since it's stabilizing but you got the idea).

**Comment by poliorcetics** [comment]
With those two commits there are now tests for this ! Still a lot of things to do though

File: src/librustdoc/html/markdown/tests.rs [hunk 1]
```diff
@@ -86,23 +88,35 @@ fn test_lang_string_parse() {
...
+    t("ignore", false, false, Ignore::All, true, false, false, false, v(), None, v());
+    t("ignore-foo", false, false, ignore_foo, true, false, false, false, v(), None, v());
+    t("should_panic", true, false, Ignore::None, true, false, false, false, v(), None, v());
+    t("no_run", false, true, Ignore::None, true, false, false, false, v(), None, v());
+    t("test_harness", false, false, Ignore::None, true, true, false, false, v(), None, v());
+    t("compile_fail", false, true, Ignore::None, true, false, true, false, v(), None, v());
+    t("allow_fail", false, false, Ignore::None, true, false, false, true, v(), None, v());
+    t("{.no_run .example}", false, true, Ignore::None, true, false, false, false, v(), None, v());
+    t("{.sh .should_panic}", true, false, Ignore::None, false, false, false, false, v(), None, v());
+    t("{.example .rust}", false, false, Ignore::None, true, false, false, false, v(), None, v());
```
Comment (jyn514) [inline src/librustdoc/html/markdown/tests.rs:104]: This is getting pretty unwieldy. Since you're modifying all the lines anyway, do you mind switching it to the builder pattern, so it's easier to read at a glance? Right now it's written for the machine to execute and only incidentally for a human to read; ideally it would be the opposite.

Comment (poliorcetics) [reply src/librustdoc/html/markdown/tests.rs:104]: Added an item to the todo-list for it, I'll do it later 😄

Comment (jyn514) [reply src/librustdoc/html/markdown/tests.rs:104]: Sounds good, yeah it doesn't need to block this new feature.

**Comment by camelid** [comment]
I saw on Zulip that you were going to call this feature `user_classes_in_docs`. Might `custom_code_classes` (maybe add `docs` somewhere, not sure if it's necessary) be a clearer name?

**Comment by poliorcetics** [comment]
It can easily be renamed as long as this isn't merged. I used a name because I needed one to proceed but I'm not particularly attached to it

**Comment by camelid** [comment]
```
tidy error: Found 1 features without a gate test.
Expected a gate test for the feature 'custom_code_classes_in_docs'.
Hint: create a failing test file named 'feature-gate-custom_code_classes_in_docs.rs'
      in the 'ui' test suite, with its failures due to
      missing usage of `#![feature(custom_code_classes_in_docs)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-custom_code_classes_in_docs line to the test file.
```

**Comment by GuillaumeGomez** [comment]
@poliorcetics Look at the PR I linked above. Everything that I removed in it is was you need to add to have a full feature.

File: src/doc/rustdoc/src/unstable-features.md [hunk 2]
```diff
@@ -38,6 +38,20 @@ future.
 Attempting to use these error numbers on stable will result in the code sample being interpreted as
 plain text.
 
+### User-provided classes in code blocks
+
+```rust
+#![feature(user_classes_in_docs)]
```
[the remark this answers is not in the document]
Comment (poliorcetics) [reply src/doc/rustdoc/src/unstable-features.md]: So that's where they go ! Thanks, I was having trouble with `src/test/ui/`

File: src/doc/rustdoc/src/unstable-features.md [hunk 3]
```diff
@@ -38,6 +38,20 @@ future.
 Attempting to use these error numbers on stable will result in the code sample being interpreted as
 plain text.
 
+### User-provided classes in code blocks
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:41]: You may want to update this since the feature name has changed:

```suggestion
### Custom CSS classes for code blocks
```

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+impl<'a, 'tcx> CustomCodeClassLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        Self { cx }
+    }
+}
+
+impl<'a, 'tcx> DocFolder for CustomCodeClassLinter<'a, 'tcx> {
+    fn fold_item(&mut self, item: Item) -> Option<Item> {
+        let cx = self.cx;
+        let dox = item.attrs.collapsed_doc_value().unwrap_or_else(String::new);
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:40]:
```suggestion
        let dox = item.attrs.collapsed_doc_value().unwrap_or_default();
```

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+impl<'a, 'tcx> CustomCodeClassLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        Self { cx }
+    }
+}
+
+impl<'a, 'tcx> DocFolder for CustomCodeClassLinter<'a, 'tcx> {
+    fn fold_item(&mut self, item: Item) -> Option<Item> {
+        let cx = self.cx;
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:39]: nit: I think you can just pass in `self.cx` directly.

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            &cx.tcx.sess.parse_sess,
+            sym::custom_code_classes_in_docs,
+            sp,
+            // This will list the wrong items to make them more easily searchable.
+            // To ensure the most correct hits, it adds back the 'class:' that was stripped.
+            &format!(
+                "Found in {} block(s): class:{}",
+                tests.nb_lang_string,
+                tests.custom_classes_found.join(", class:")
+            ),
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:88]: I find this error pretty confusing. What about this instead?
```suggestion
            "custom classes in code blocks are unstable",
       ).note(
            // This will list the wrong items to make them more easily searchable.
            // To ensure the most correct hits, it adds back the 'class:' that was stripped.
            &format!(
                "found in {} block(s): class:{}",
                tests.nb_lang_string,
                tests.custom_classes_found.join(", class:")
            ),
```

File: src/tools/tidy/src/features.rs [hunk 7]
```diff
@@ -89,6 +89,15 @@ pub fn check(
             &src_path.join("test/ui"),
             &src_path.join("test/ui-fulldeps"),
             &src_path.join("test/compile-fail"),
+            // This allows `tidy` to pick up features that are `rustdoc`-specific.
+            // They won't be run by `tidy` but they will be run later in the
+            // process when `rustdoc` is ran on the files in `rustdoc-ui`.
```
Comment (jyn514) [inline src/tools/tidy/src/features.rs:94]:
```suggestion
            // test suite when `rustdoc` is run on the files in `rustdoc-ui`.
```

File: src/tools/tidy/src/features.rs [hunk 8]
```diff
@@ -89,6 +89,15 @@ pub fn check(
             &src_path.join("test/ui"),
             &src_path.join("test/ui-fulldeps"),
             &src_path.join("test/compile-fail"),
+            // This allows `tidy` to pick up features that are `rustdoc`-specific.
+            // They won't be run by `tidy` but they will be run later in the
```
Comment (jyn514) [inline src/tools/tidy/src/features.rs:93]: I'm confused what it means to be 'run by `tidy`'. Is there something different between rustdoc-ui and the other tests?

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 6 above)
Comment (camelid) [reply src/librustdoc/passes/check_custom_code_classes.rs:88]: I don't think we should list how many we found – we're pointing using the span. I think the error should say something like `found custom class on code block: class:some-custom-class`.

File: src/tools/tidy/src/features.rs (hunk 8 above)
Comment (poliorcetics) [reply src/tools/tidy/src/features.rs:93]: I was tired when I wrote that. I'll fix it in the next commit

**Comment by poliorcetics** [comment]
Barring errors in CI, this should now only need clear up work !

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 9]
```diff
@@ -79,11 +74,13 @@ crate fn look_for_custom_classes<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &
...
             sp,
+            "custom classes in code blocks are unstable",
+        )
+        .note(
             // This will list the wrong items to make them more easily searchable.
             // To ensure the most correct hits, it adds back the 'class:' that was stripped.
             &format!(
-                "Found in {} block(s): class:{}",
-                tests.nb_lang_string,
+                "Found custom class(es) on code block(s): class:{}",
```
Comment (camelid) [inline src/librustdoc/passes/check_custom_code_classes.rs:83]: Rust's general convention is to have lowercase errors/notes/etc., so this should be:

```suggestion
                "found custom class(es) on code block(s): class:{}",
```

But also I feel that the pluralization in the parentheses (whatever the name for that is!) is somewhat awkward. I recommend something like this:

```suggestion
                "found these custom classes: class:{}",
```

which follows the style of some other Rust errors.

Comment (poliorcetics) [reply src/librustdoc/passes/check_custom_code_classes.rs]: Changed for what you recommended. 😄

**Comment by poliorcetics** [comment]
Only the tests to clean up now (https://github.com/rust-lang/rust/pull/79454#discussion_r531814290)

**Comment by GuillaumeGomez** [comment]
You need to run `cargo fmt` on `librustdoc` (nightly version).

**Comment by camelid** [comment]
Not `./x.py fmt`?

**Comment by GuillaumeGomez** [comment]
Works too I guess? Never used it. :laughing:

**Comment by poliorcetics** [comment]
I have done everything that was in my todo-list for this PR, it's ready for a more in-depth review.

File: src/librustdoc/html/markdown/tests.rs [hunk 10]
```diff
@@ -51,82 +51,97 @@ fn test_unique_id() {
...
-                compile_fail,
-                error_codes,
-                original: s.to_owned(),
-                allow_fail,
-                edition,
-            }
-        )
+    fn t(lg: LangString) {
+        let s = &lg.original;
+        assert_eq!(LangString::parse(s, ErrorCodes::Yes, true, None), lg)
```
Comment (jyn514) [inline src/librustdoc/html/markdown/tests.rs:56]: Do you mind separating this into a separate PR? The current one is still waiting on the FCP in https://github.com/rust-lang/rust/pull/78429 and I'd love to have the improvement here early :)

**Review by jyn514 — COMMENTED** [review]
Didn't review in super detail, but overall this looks good. I like that the syntax makes this easy to add to the JSON backend if we want :)

File: src/librustdoc/html/markdown/tests.rs (hunk 10 above)
Comment (poliorcetics) [reply src/librustdoc/html/markdown/tests.rs:56]: Will do !

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks all good to me, thanks! If no one else has fixes in mind, I think we're good to go (maybe a squash beforehand? 23 commits seems a lot in here...)

ping @jyn514

**Comment by Hywan** [comment]
:+1:

**Comment by jyn514** [comment]
@GuillaumeGomez this is still waiting on the FCP in https://github.com/rust-lang/rust/pull/78429

**Comment by poliorcetics** [comment]
Could this be merged even if #78429 is not ? With all the work happening on refactoring rustdoc, rebase are complicated and they become harder as time passes.

**Comment by GuillaumeGomez** [comment]
I wonder if you didn't fail the last rebase a bit because "Move tooltip messages ou of html" is a commit of mine. :p

Also, sorry but #78429 needs to be merged first. What I'd recommend though to make your life simpler would be to make your git history shorter. If you feel too overcome for this, I'll make next rebase before merging. Don't worry, it'll get in at some point. Just a lot of things are going on and it's kinda the worst time of the year for contributions. :laughing: But be sure that your work is strongly appreciated and that we can't wait to have it merged!

**Comment by camelid** [comment]
If you want to undo your rebase and try again (perhaps squashing first to make it easier), I think you can just `git reset --hard 3a46f97` which is the commit GitHub shows as the old branch tip.

Also, I find `git range-diff master @{u} HEAD` to be helpful sometimes when I've just done a rebase but haven't pushed it up yet. It sort of diffs each commit instead of the total changes (at least I think that's what it does!).

**Comment by poliorcetics** [comment]
The rebase succeeded, but the resulting code didn't compile so I had to fix some things.

> make your git history shorter

I'll do that, I should have thought about it.

File: compiler/rustc_feature/src/active.rs [hunk 11]
```diff
@@ -620,6 +620,9 @@ declare_features! (
     /// Allows arbitrary expressions in key-value attributes at parse time.
     (active, extended_key_value_attributes, "1.50.0", Some(78835), None),
 
+    /// Allows users to provide classes for fenced code block using `class:classname`.
+    (active, custom_code_classes_in_docs, "1.50.0", Some(79483), None),
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:624]: I don't think this will land in 1.50, so:

```suggestion
    (active, custom_code_classes_in_docs, "1.51.0", Some(79483), None),
```

File: src/doc/rustdoc/src/unstable-features.md [hunk 12]
```diff
@@ -38,6 +38,20 @@ future.
...
+```rust
+#![feature(custom_code_classes_in_docs)]
+
+/// ```class:language-c
+/// int main(void) { return 0; }
+/// ```
+fn main() {}
+```
+
+The text `int main(void) { return 0; }` is copied without formatting in a code block with the class
```
Comment (camelid) [inline src/doc/rustdoc/src/unstable-features.md:52]: What does "without formatting" mean?

File: src/librustdoc/html/markdown.rs [hunk 13]
```diff
@@ -283,28 +285,36 @@ impl<'a, I: Iterator<Item = Event<'a>>> Iterator for CodeBlocks<'_, 'a, I> {
...
+        } else {
+            let classes = if special_class.is_empty() {
+                added_classes.join(" ")
+            } else {
+                format!("{} {}", special_class.trim(), added_classes.join(" "))
+            };
+
+            s.push_str(&highlight::render_with_added_classes(text, classes, tooltip));
+        }
+
```
Comment (camelid) [inline src/librustdoc/html/markdown.rs:317]: Isn't this from that unrelated commit by Guillaume (cc https://github.com/rust-lang/rust/pull/79454/#issuecomment-751829416)?

File: src/tools/tidy/src/features.rs [hunk 14]
```diff
@@ -89,6 +89,13 @@ pub fn check(
             &src_path.join("test/ui"),
             &src_path.join("test/ui-fulldeps"),
             &src_path.join("test/compile-fail"),
+            // This allows `tidy` to pick up features that are `rustdoc`-specific.
+            //
+            // This way, `tidy` is happy because all features have a test file
+            // and all the features are correctly tested (either through `rustdoc`
+            // or `rustc`), while avoiding having  to run `rustdoc` for all other
```
Comment (camelid) [inline src/tools/tidy/src/features.rs:96]: nit: there's an extra space here

```suggestion
            // or `rustc`), while avoiding having to run `rustdoc` for all other
```

File: src/librustdoc/html/markdown.rs (hunk 13 above)
Comment (poliorcetics) [reply src/librustdoc/html/markdown.rs:317]: The removed parts are, yes, but the parts in green are from me (they notably add the support for the custom classes to the rendered html)

File: src/doc/rustdoc/src/unstable-features.md (hunk 12 above)
Comment (poliorcetics) [reply src/doc/rustdoc/src/unstable-features.md]: I changed for `without highlighting`, is it clearer ?

Comment (camelid) [reply src/doc/rustdoc/src/unstable-features.md]: Yes, that's better! I think "The text is rendered without highlighting" (not "copied") sounds better, but that's not a big deal.

Comment (poliorcetics) [reply src/doc/rustdoc/src/unstable-features.md]: I changed for `rendered` while fixing the conflicts, thanks for the catch !

**Comment by poliorcetics** [comment]
Rebased and updated to 1.52.0 for the feature

File: compiler/rustc_feature/src/active.rs [hunk 15]
```diff
@@ -126,7 +126,7 @@ declare_features! (
     /// macros disappear).
     (active, allow_internal_unsafe, "1.0.0", None, None),
 
-    /// no-tracking-issue-end
+    // no-tracking-issue-end
```
Comment (jyn514) [inline compiler/rustc_feature/src/active.rs:129]: Why did this change?

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 16]
```diff
@@ -384,7 +384,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ///
     /// InferBorrowKind results in a structure like this:
     ///
-    /// ```
+    /// ```text
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/upvar.rs:387]: This seems like it should be part of https://github.com/rust-lang/rust/pull/79816. Is there a reason it's also here?

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+crate fn look_for_custom_classes<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &Item) {
+    let hir_id = match cx.as_local_hir_id(item.def_id) {
+        Some(hir_id) => hir_id,
+        None => {
+            // If non-local, no need to check anything.
+            return;
+        }
+    };
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:64]: Hmm, I guess this is a policy question: if you re-export an item using `feature`s, should you also need to enable the feature in your own crate? @Manishearth what do you think?

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+    if !tests.custom_classes_found.is_empty() && !cx.tcx.features().custom_code_classes_in_docs {
+        debug!("reporting error for {:?} (hid_id={:?})", item, hir_id);
+        let sp = span_of_attrs(&item.attrs).unwrap_or(item.source.span());
+        feature_err(
+            &cx.tcx.sess.parse_sess,
+            sym::custom_code_classes_in_docs,
+            sp,
+            "custom classes in code blocks are unstable",
+        )
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:78]: This is awesome, thank you :heart:

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+use crate::fold::DocFolder;
+use crate::html::markdown::{find_testable_code, ErrorCodes, LangString};
+
+use rustc_session::parse::feature_err;
+use rustc_span::symbol::sym;
+
+crate const CHECK_CUSTOM_CODE_CLASSES: Pass = Pass {
+    name: "check-custom-code-classes",
+    run: check_custom_code_classes,
+    description: "check for custom code classes while not in nightly",
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:18]:
```suggestion
    description: "check for custom code classes without a feature-gate enabled",
```

**Review by jyn514 — COMMENTED** [review]
Thought I'd start on reviewing the implementation while we wait for the FCP.

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 17 above)
Comment (Manishearth) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: Feels like it would be very annoying for reexports, especially since this is docs-only. We can, however, _not render it_ if the feature isn't enabled.

Comment (jyn514) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: > We can, however, not render it if the feature isn't enabled.

Oh, I don't know if I like that. That means the documentation changes just by re-exporting it.

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 19 above)
Comment (camelid) [reply src/librustdoc/passes/check_custom_code_classes.rs:18]: Or, since it has to be a specific feature gate:

```suggestion
    description: "check for custom code classes without the feature-gate enabled",
```

File: compiler/rustc_feature/src/active.rs (hunk 15 above)
Comment (camelid) [reply compiler/rustc_feature/src/active.rs:129]: It was a bug in the original code. Currently [the docs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_feature/struct.Features.html#structfield.link_llvm_intrinsics) look like:

image

because the doc comment is processed as being attached to `link_llvm_intrinsics`.

File: compiler/rustc_typeck/src/check/upvar.rs (hunk 16 above)
Comment (poliorcetics) [reply compiler/rustc_typeck/src/check/upvar.rs:387]: That may have been me mixing up two compilers while switching branches and not paying attention, I'll change it back

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 19 above)
Comment (poliorcetics) [reply src/librustdoc/passes/check_custom_code_classes.rs:18]: I'm suddenly wondering if the negative here is correct. Maybe it should be `with the feature-gate enabled` ? I don't remember in which context this message pops up.

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 17 above)
Comment (poliorcetics) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: I'm not sure how to resolve that, I mostly copied what was done in other passes when ignoring external items.

Comment (jyn514) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: > @jyn514 Does it? I assume the custom classes are only used by additional stylesheets, and reexports wouldn't have those stylesheets. So to me this is already broken.

So, Guillaume said

> The point here is to allow users to add a class themselves and disable the rustdoc highlighting so that they can then add a JS library which will do it.

It's unclear to me how people will actually be adding their own highlighting. Is the idea to use `--extend-css` for this? That would work regardless of the crate it gets re-exported in if you forget to pass it; and would work regardless of the crate if the class gets added.

> I'm not sure how to resolve that, I mostly copied what was done in other passes when ignoring external items.

You only use the HIR ID for lints, so depending on the policy we want you could either not warn at all, or warn on the `pub use` instead of the original item.

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 19 above)
Comment (jyn514) [reply src/librustdoc/passes/check_custom_code_classes.rs:18]: This doesn't pop up anywhere except `-Z time-passes` I think. Since this pass is explicitly only emitting lints and not doing anything else, I think @camelid's suggestion is fine.

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 17 above)
Comment (GuillaumeGomez) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: Either `--extend-css` or `--html-in-header` and then pointing to hightlight.js or equivalent.

**Review by Deloper2297 — APPROVED** [review]

**Comment by camelid** [comment]
No longer blocked! #78429 has been merged. 🎉 

@rustbot label: -S-blocked

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 17 above)
Comment (jyn514) [reply src/librustdoc/passes/check_custom_code_classes.rs:64]: Ok, so I think we should unconditionally apply the custom class, and not check the feature gate (because that breaks crates: https://github.com/rust-lang/rust/issues/82284).

File: src/librustdoc/passes/check_custom_code_classes.rs (hunk 19 above)
Comment (jyn514) [reply src/librustdoc/passes/check_custom_code_classes.rs:18]: @poliorcetics this is still waiting on you to accept the change I think.

File: src/librustdoc/passes/check_custom_code_classes.rs [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+struct CustomCodeClassLinter<'a, 'tcx> {
+    cx: &'a DocContext<'tcx>,
+}
+
+impl<'a, 'tcx> CustomCodeClassLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        Self { cx }
+    }
+}
```
Comment (jyn514) [inline src/librustdoc/passes/check_custom_code_classes.rs:35]: This seems like a lot of boilerplate for 8 characters, I would just use `CustomCodeClassLinter { cx }` in `fn check_custom_code_classes`.
```suggestion
```

File: src/librustdoc/html/markdown.rs [hunk 21]
```diff
@@ -868,6 +881,14 @@ impl LangString {
                         seen_other_tags = true;
                     }
                 }
+                x if x.starts_with("class:") => {
+                    let class = x.trim_start_matches("class:");
+                    if class.is_empty() {
+                        seen_other_tags = true;
```
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:887]: Hmm, what does this do? I would expect `class:` to give an error (or at least a warning).

**Review by jyn514 — COMMENTED** [review]
This is still waiting on https://github.com/rust-lang/rust/pull/79454#discussion_r588906949 and https://github.com/rust-lang/rust/pull/79454#discussion_r575912125 (although I don't care very much about the second one).

**Comment by crlf0710** [comment]
@poliorcetics Ping from triage! Could you address the comments above? Thanks!

**Comment by jyn514** [comment]
Closing due to inactivity.
